### PR TITLE
KV Store: Graveler implement Reset

### PR DESCRIPTION
**Pull request by N-o-Z**
Decided to split the PRs eventually

### CODE REVIEW

File: pkg/graveler/graveler.go [hunk 1]
```diff
@@ -1851,8 +1846,51 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
...
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+	currBranch, err := g.RefManager.GetBranch(ctx, repositoryID, branchID)
+	if err != nil {
+		return err
+	}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1876]: Use the branch given to you by the `BranchUpdate` func. The value here may be outdated.
```suggestion
```

File: pkg/graveler/graveler.go [hunk 2]
```diff
@@ -1851,8 +1846,51 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
...
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		if currBranch.StagingToken != branch.StagingToken { // Check staging token hasn't changed after get with predicate
+			return nil, ErrConflictFound
+		}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1881]: `BranchUpdate` got your back! This check is redundant

File: pkg/graveler/graveler.go [hunk 3]
```diff
@@ -1851,8 +1846,51 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
...
+		// Zero tokens and try to set branch
+		branch.StagingToken = generateStagingToken(repositoryID, branchID)
+		branch.SealedTokens = make([]StagingToken, 0)
+		return branch, nil
+	})
+
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1887]: remove line

File: pkg/graveler/graveler.go (hunk 2 above)
Comment (itaiad200) [reply pkg/graveler/graveler.go:1881]: Just save the old tokens in a list

File: pkg/graveler/graveler.go [hunk 4]
```diff
@@ -1851,8 +1846,51 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
...
 
+// dropStaging deletes all staging area entries of a given branch from store
+// We do not wait for result as this is an asynchronous operation
+func (g *KVGraveler) dropStaging(ctx context.Context, branch *Branch) {
+	tokens := branch.SealedTokens
+	tokens = append(tokens, branch.StagingToken)
+	for _, st := range tokens {
+		token := st // pinning
+		go func() {
+			err := g.StagingManager.Drop(ctx, token)
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1857]: I think the error should be returned, why shadow it from the caller?

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks - dropped some comments. Also - I know we don't have tests before, please consider adding a single sanity test to the graveler `Reset` functionality.

File: pkg/graveler/graveler.go (hunk 4 above)
Comment (N-o-Z) [reply pkg/graveler/graveler.go:1857]: Once we update the branch for all things considered we've reset the branch. Waiting for the error to be returned will prolong the operation and return an error which is not interesting for the user

File: pkg/graveler/graveler.go [hunk 5]
```diff
@@ -1851,8 +1846,45 @@ func (g *KVGraveler) stagingEmpty(ctx context.Context, repositoryID RepositoryID
...
 
+// dropStaging deletes all staging area entries of a given branch from store
+// We do not wait for result as this is an asynchronous operation
+func (g *KVGraveler) dropTokens(ctx context.Context, tokens []StagingToken) {
+	for _, st := range tokens {
+		token := st // pinning
+		go func() {
+			err := g.StagingManager.Drop(ctx, token)
+			if err != nil {
+				logging.Default().WithError(err).WithField("staging_token", token).Error("failed to drop staging token")
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1857]:
```suggestion
				logging.Default().WithError(err).WithField("staging_token", token).Error("Failed to drop staging token")
```

**Comment by N-o-Z** [comment]
> Thanks - dropped some comments. Also - I know we don't have tests before, please consider adding a single sanity test to the graveler `Reset` functionality.

@itaiad200 I don't really know how we can unit test this in the graveler context. The whole branch update and drop logics are mocked.
It is covered though by system tests

**Comment by N-o-Z** [comment]
Found a bug in Get flow

File: pkg/graveler/graveler.go [hunk 6]
```diff
@@ -1335,14 +1335,17 @@ func (g *KVGraveler) Get(ctx context.Context, repositoryID RepositoryID, ref Ref
...
+		if err != nil {
+			if !errors.Is(err, ErrNotFound) {
+				return nil, err
+			} // proceed to committed
+		} else { // In staging
+			if value == nil {
+				// tombstone - the entry was deleted on the branch => doesn't exist
+				return nil, ErrNotFound
+			}
+			return value, nil
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1347]: Great catch!!

I think it deserves its own PR. Also, there are tests for `Get` flow in Graveler. Can we add/change them to fit this scenario too?

Comment (N-o-Z) [reply pkg/graveler/graveler.go:1347]: The tests were there, they were just outdated. Unfortunately I suspect we have many such tests where the flow is not tested correctly due to mock configurations. We should take action once we remove the DB support to refactor the tests and clean up the mocks

**Review by itaiad200 — APPROVED** [review]
LGTM
